File: libsignal/service/src/test/java/org/whispersystems/signalservice/api/services/DonationsServiceTest.java
package org.whispersystems.signalservice.api.services;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;
import org.mockito.Mockito;
import org.whispersystems.signalservice.api.push.exceptions.NonSuccessfulResponseCodeException;
import org.whispersystems.signalservice.api.subscriptions.ActiveSubscription;
import org.whispersystems.signalservice.api.subscriptions.SubscriberId;
import org.whispersystems.signalservice.internal.ServiceResponse;
import org.whispersystems.signalservice.internal.push.PushServiceSocket;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(JUnit4.class)
public class DonationsServiceTest {

  private final PushServiceSocket pushServiceSocket = Mockito.mock(PushServiceSocket.class);
  private final DonationsService  testSubject       = new DonationsService(pushServiceSocket);

  @Test
  public void givenASubscriberId_whenIGetASuccessfulResponse_thenItIsMappedWithTheCorrectStatusCodeAndNonEmptyObject() throws Exception {
    // GIVEN
    SubscriberId subscriberId = SubscriberId.generate();
    when(pushServiceSocket.getSubscription(subscriberId.serialize()))
        .thenReturn(getActiveSubscription());

    // WHEN
    ServiceResponse<ActiveSubscription> response = testSubject.getSubscription(subscriberId);

    // THEN
    verify(pushServiceSocket).getSubscription(subscriberId.serialize());
    assertEquals(200, response.getStatus());
    assertTrue(response.getResult().isPresent());
  }

  @Test
  public void givenASubscriberId_whenIGetAnUnsuccessfulResponse_thenItIsMappedWithTheCorrectStatusCodeAndEmptyObject() throws Exception {
    // GIVEN
    SubscriberId subscriberId = SubscriberId.generate();
    when(pushServiceSocket.getSubscription(subscriberId.serialize()))
        .thenThrow(new NonSuccessfulResponseCodeException(403));

    // WHEN
    ServiceResponse<ActiveSubscription> response = testSubject.getSubscription(subscriberId);

    // THEN
    verify(pushServiceSocket).getSubscription(subscriberId.serialize());
    assertEquals(403, response.getStatus());
    assertFalse(response.getResult().isPresent());
  }

  private ActiveSubscription getActiveSubscription() {
    return ActiveSubscription.EMPTY;
  }
}

File: libsignal-service/src/test/java/org/whispersystems/signalservice/api/services/DonationsServiceTest.kt
package org.whispersystems.signalservice.api.services

import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import org.whispersystems.signalservice.api.NetworkResult
import org.whispersystems.signalservice.api.donations.DonationsApi
import org.whispersystems.signalservice.api.push.exceptions.NonSuccessfulResponseCodeException
import org.whispersystems.signalservice.api.subscriptions.ActiveSubscription
import org.whispersystems.signalservice.api.subscriptions.SubscriberId

class DonationsServiceTest {
  private val donationsApi: DonationsApi = mockk<DonationsApi>()
  private val testSubject = DonationsService(donationsApi)
  private val activeSubscription = ActiveSubscription.EMPTY

  @Test
  fun givenASubscriberId_whenIGetASuccessfulResponse_thenItIsMappedWithTheCorrectStatusCodeAndNonEmptyObject() {
    // GIVEN
    val subscriberId = SubscriberId.generate()
    every { donationsApi.getSubscription(subscriberId) } returns NetworkResult.Success(activeSubscription)

    // WHEN
    val response = testSubject.getSubscription(subscriberId)

    // THEN
    verify { donationsApi.getSubscription(subscriberId) }
    assertEquals(200, response.status)
    assertTrue(response.result.isPresent)
  }

  @Test
  fun givenASubscriberId_whenIGetAnUnsuccessfulResponse_thenItIsMappedWithTheCorrectStatusCodeAndEmptyObject() {
    // GIVEN
    val subscriberId = SubscriberId.generate()
    every { donationsApi.getSubscription(subscriberId) } returns NetworkResult.StatusCodeError(NonSuccessfulResponseCodeException(403))

    // WHEN
    val response = testSubject.getSubscription(subscriberId)

    // THEN
    verify { donationsApi.getSubscription(subscriberId) }
    assertEquals(403, response.status)
    assertFalse(response.result.isPresent)
  }
}
